Доработана функциональность по техническому заданию спринта

Omorlan opened the pull request:
feat: Добавить обработку сетевых запросов
feat: Добавить тесты под новый функционал

Omorlan committed:
feat: Добавить обработку сетевых запросов
feat: Добавить тесты под новый функционал

Omorlan committed:
fix: CodeStyle

Omorlan commented:
![review](https://github.com/Omorlan/java-kanban/assets/140850406/b033ceb9-1f42-4346-960d-86077d4cf9a2)

Omorlan committed:
fix: Исправить замечания SonarList
fix: Провести рефакторинг

afomkina commented on src/server/adapter/LocalDateTimeAdapter.java:
```diff
@@ -0,0 +1,35 @@
+import java.time.format.DateTimeFormatter;
+
+public class LocalDateTimeAdapter extends TypeAdapter<LocalDateTime> {
+    private final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("dd.MM.yyy HH:mm");
```
Лучше вынести паттерн в отдельный утильный класс и сделать константой, чтобы можно было переиспользовать его в нескольких классах

afomkina commented on test/src/server/HttpTaskServerTest.java:
```diff
@@ -0,0 +1,269 @@
+
+class HttpTaskServerTest {
+
+    HttpTaskServer taskServer;
```
У некоторых полей класса не хватает модификатора доступа

afomkina commented on test/src/server/HttpTaskServerTest.java:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    /*
+    POST
```
Предлагаю немного подекомпозировать тесты и разбить из на логические составляющие. Это можно сделать с помощью [@Nested классов](https://www.petrikainulainen.net/programming/testing/junit-5-tutorial-writing-nested-tests/). Так код станет более читаемым и его будет проще поддерживать

Omorlan replied:
надеюсь я правильно понял концепцию

afomkina commented on src/server/handler/EpicsHandler.java:
```diff
@@ -0,0 +1,99 @@
+                    break;
+            }
+        } catch (Exception exception) {
+            exception.printStackTrace();
```
Лучше не печатать стектрейс, а логировать мессадж ошибки:
`System.out.print(exception.getMessage());`
Чаще всего весь стектрейс нам не нужен(он большой и малоинформативный), а достаточного всего лишь сообщения

afomkina commented on src/server/handler/HistoryHandler.java:
```diff
@@ -0,0 +1,32 @@
+                exchange.sendResponseHeaders(500, 0);
+            }
+        } catch (Exception exception) {
+            exception.printStackTrace();
```
Лучше просто распечатать мессадж исключения

afomkina commented on src/server/handler/PrioritizedHandler.java:
```diff
@@ -0,0 +1,33 @@
+                exchange.sendResponseHeaders(500, 0);
+            }
+        } catch (Exception exception) {
+            exception.printStackTrace();
```
И здесь

afomkina commented on src/server/handler/SubtasksHandler.java:
```diff
@@ -0,0 +1,126 @@
+
+            }
+        } catch (Exception exception) {
+            exception.printStackTrace();
```
И здесь

afomkina commented on src/server/handler/SubtasksHandler.java:
```diff
@@ -0,0 +1,126 @@
+                            exchange.sendResponseHeaders(500, 0);
+                        }
+                    } else if (Pattern.matches("^/subtasks/\\d+$", path)) {
+                        String pathId = path.replaceFirst(SUBTASKS, "");
```
Код со строк 91-99 повторяется в классе `TasksHandler` на строках 90-98. Нужно выделить его в отдельный метод  с сигнатурой
`protected String yourName(String path, String url, HttpExchange exchange) throws IOException`
в классе `Handler` и переиспользовать, а не дублировать

afomkina commented on src/server/handler/SubtasksHandler.java:
```diff
@@ -0,0 +1,126 @@
+                        String pathId = path.replaceFirst(SUBTASKS, "");
+                        int id;
+                        try {
+                            id = parsePathId(pathId);
```
Переменная id нигде не используется, ее нужно либо использовать, либо удалить

Omorlan replied:
убрав id, пропала и проблема с дублированием

afomkina commented on src/server/handler/SubtasksHandler.java:
```diff
@@ -0,0 +1,126 @@
+                            response = gson.toJson(subtask);
+                            sendResponse(exchange, response);
+                        } else {
+                            exchange.sendResponseHeaders(500, 0);
```
Задание со `"*"`:
Можно создать отдельное перечисление для HTTP кодов и использовать его вместо "голых" цифр. 
Так код будет выглядеть более читаемо и понятней. 
Примеры можно посмотреть в главе [Конструкторы, поля и методы перечисления](https://metanit.com/java/tutorial/3.8.php#:~:text=%D0%9A%D0%BE%D0%BD%D1%81%D1%82%D1%80%D1%83%D0%BA%D1%82%D0%BE%D1%80%D1%8B%2C%20%D0%BF%D0%BE%D0%BB%D1%8F%20%D0%B8%20%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D1%8B%20%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%81%D0%BB%D0%B5%D0%BD%D0%B8%D1%8F)

Omorlan replied:
чтож я был готов и к этому
![изображение_2024-03-31_193317662](https://github.com/Omorlan/java-kanban/assets/140850406/ab075b6e-cf74-40cb-9e4c-5895e30b430d)

Omorlan committed:
fix: Вынести паттерн времени в константу
fix: Добавить недостающий модификаторы доступы
fix: Улучшить читаемость тестов
fix: Логировать ошибки
fix: Убрать неиспользуеммые перменные
fix: Добавить перечисление для кодов

afomkina commented on src/resources/HttpStatusCode.java:
```diff
@@ -0,0 +1,20 @@
+package resources;
+
+public enum HttpStatusCode {
```
🔥

afomkina commented on src/resources/UtilConstant.java:
```diff
@@ -0,0 +1,5 @@
+package resources;
+
+public class UtilConstant {
```
🔥

afomkina reviewed (approved):
Получилось здорово, молодец 🔥
